### PR TITLE
Deprecate InstanceOperations methods that take a String for server

### PR DESCRIPTION
**Pull request by dlmarion**
Created ServerType API objects that contain the server type and address. Deprecated API methods on InstanceOperations that take a String for the server and created replacement methods. Related to #3593.

### CODE REVIEW

**Comment by EdColeman** [comment]
Should the javadocs include the `@deprecated` annotation along with the use @link suggestion?

**Comment by dlmarion** [comment]
> Should the javadocs include the `@deprecated` annotation along with the use @link suggestion?

I can add that. I'm going to wait to see if I get other comments before making this change.

File: core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java:220]: It would be nice to have something like the following, not sure if the following javadoc syntax is correct.

```suggestion
   * @throws IllegalArgumentException when the type of server is not SCAN_SERVER or TABLET_SERVER
   * @since 3.0.0
```

File: core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java:222]: Unless there is a specific reason to use list, using a more generic type gives more flexibility w/ changing the impl

```suggestion
  Collection<ActiveScan> getActiveScans(Server server)
```

File: core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java:194]: Wondering how this method could also return information about a servers group.  Was thinking through the following possible options.

 1.  Group info could be added to the Sever class.
 2. This method could return something like Map<Server, ServerInfo> where ServerInfo includes the group
 3. Could add a new method to `String getGroup(Server server)` to InstanceOperations.

File: core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java:177]: Somehow information can be added to the deprecation annotation that indicates the version where the deprecation happened.

File: core/src/main/java/org/apache/accumulo/core/client/admin/servers/Server.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/client/admin/servers/Server.java:57]: Why handle nulls here?

File: core/src/main/java/org/apache/accumulo/core/client/admin/servers/Server.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/client/admin/servers/Server.java:82]: The constructor does not allow null, so do not need to check for it here.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/InstanceOperationsImpl.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/InstanceOperationsImpl.java:254]: ExternalCompactionUtil.getCompactorAddrs is not public API, should not recommend calling it from public API method.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/InstanceOperationsImpl.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/InstanceOperationsImpl.java:446]: Would be a follow on issue, would be nice to make this work w/ all the serer types.

File: core/src/main/java/org/apache/accumulo/core/summary/Gatherer.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/summary/Gatherer.java:163]: Could this use the new method and avoid the suppression?

File: core/src/main/java/org/apache/accumulo/core/client/admin/servers/Server.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/client/admin/servers/Server.java:26]: Wonder if this should have a more specific name.  Could be ServerAddress or AccumuloServerAddress.  If the group info were to be included in this class, then maybe AccumuloServer or AccumuloServerInfo

File: core/src/main/java/org/apache/accumulo/core/client/admin/servers/Server.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/client/admin/servers/Server.java:82]: This was code generated by the IDE. I can fix.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/InstanceOperationsImpl.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/InstanceOperationsImpl.java]: Removed this exception in ea67851. Since I added ResourceGroup to Server, I went ahead and implemented this case.

File: core/src/main/java/org/apache/accumulo/core/client/admin/servers/Server.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/client/admin/servers/Server.java]: Removed null checks in ea67851.

File: core/src/main/java/org/apache/accumulo/core/client/admin/servers/Server.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/client/admin/servers/Server.java]: I removed these checks and implemented something different in ea67851.  If first or second is null, then I throw an NPE, which is documented in [Comparator](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Comparator.html#compare(T,T))

File: core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java]: I added `resourceGroup` to `Server` in ea67851. This is a little before it's time, as this is fully implemented in the elasticity branch.

File: core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java]: Implemented in ea67851.

File: core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java:225]: Added in in ea67851.

File: core/src/main/java/org/apache/accumulo/core/client/admin/servers/ServerType.java
Comment (dlmarion) [inline core/src/main/java/org/apache/accumulo/core/client/admin/servers/ServerType.java:22]: Note that I did not add `GC` or `MONITOR` here as the client cannot interact with them directly.

File: core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java
Comment (dlmarion) [inline core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java:249]: This needs to be updated.

File: core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java
Comment (dlmarion) [inline core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java:221]: This needs to be updated.

File: core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java]: > This is a little before it's time, as this is fully implemented in the elasticity branch.

What is the diff between main and elasticity that impacts this PR?

Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java]: My comment is that the name `resourceGroup` is a little before it's time. In 2.1 ScanServers have groups and Compactors have queues, which is what will be returned here for 3.0. In elasticity as you know, we have centralized on the name `group` in the properties and code, and I have been calling the feature `resource groups` as `groups` is too vague and the way that they are intended to be used is to create groups of resources dedicated to running tservers/sservers/compactors for specific use-cases.

Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java]: May make sense to defer adding the resourceGroup to the server type to the elasticity branch.  Could still add these new types in 3.0 and not have the resource group method in 3.0.  Thinking if we release something in 3.x that still may still change in the elasticity branch that it could cause problems if it does change.

Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java]: @ctubbsii  said he wanted to review this too. I'll wait for his feedback and make this changes along with anything that he suggests.

File: core/src/main/java/org/apache/accumulo/core/client/admin/servers/Server.java
Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/client/admin/servers/Server.java:26]: This seems more like a ServerId rather than a Server object. It contains identifying information about a server, which may or may not exist.

This could extend AbstractId, and be constructed similarly to those, to avoid creating redundant instances that refer to the same server.

This should also not implement `Comparator<Server>`, it should be something like `Server<T extends Server<?>> implements Comparable<T>`, so that 1) it's a Comparable object, not a Comparator, and 2) it is only comparable to instances of the same type.

File: core/src/main/java/org/apache/accumulo/core/client/admin/servers/ServerType.java
Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/client/admin/servers/ServerType.java:22]: Why is this one public API? It seems the enum is only used internally for the various Server subclasses. Do we even need this enum at all?

File: core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java:227]: You could enforce this in the API contract at compile time by having an intermediate ScanningServer type, rather than leave this to a runtime failure.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/InstanceOperationsImpl.java
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/InstanceOperationsImpl.java:256]: I see this is where the type enum is used by users. But, you could just as easily get rid of the enum, and users can pass a `Predicate<Server>`, which is far more flexible, and they can do things like `getServers(s -> s.getHost().startsWith("somePrefix"));` or `getServers(s -> s instanceof TabletServer)`.

**Review by ctubbsii — CHANGES_REQUESTED** [review]
Overall, I think this is nice stuff, but since this is new API stuff, I think we should wait until we cut 3.0.0 (which I plan to do an RC for soon), and drop this in 3.1.0, with some additional polishing.

I also think there should be a difference between the references to servers, and a server name/id. I think we should have a ServerId type that extends AbstractId, and a separate Server type that is a "first-class citizen" in the API that represents a reference to the Server, with operations methods dangling off of it. This would mesh really well with the previous ideas I've had about doing the same thing with concrete Namespace and Table objects in the API, which I would also like to work on for 3.1.

Also, I think most of the places where there are warnings suppressions can be updated to use the newer types, and we just have a single place where we test that the deprecated methods still return what we expect (i.e. return the same thing as the new methods). That way, it's easier to remove the deprecated ones later, since our own code wouldn't be using it anymore. It would also provide a way of testing to the new APIs to make sure it is fully functional for a wide variety of use cases.

I'm also interested in how we'd want to do permissions on these APIs... since I think a lot of this is just looking at ZK as a client, there may not really be any authentication checks to get a list of the servers. A lot of these changes are superficial API changes, though, so I'd have to dig a bit more to see what's being done with the current API. Maybe there's an opportunity to improve the security of these APIs during this transition, instead of merely making superficial changes to the API types to de-stringify them.

File: core/src/main/java/org/apache/accumulo/core/client/admin/servers/ManagerServer.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/client/admin/servers/ManagerServer.java:25]: If not specified, there is a default resource group. So while it's not used, there is one, it's the default.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/InstanceOperationsImpl.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/InstanceOperationsImpl.java:256]: If I use the Predicate then I have to test against every single server each time this is called, right? This is replacing `getTabletServers`, `getScanServers` and `getManagerLocations`. If the user want's everything, then I think it makes sense that they call this method N times instead of us gathering all of the information every time.

Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/InstanceOperationsImpl.java:256]: Ok, I think I figured out a solution for this.

Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/InstanceOperationsImpl.java]: Solution implemented in f63dce0

File: core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java]: Resolved in f63dce0

File: core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java]: Resolved in f63dce0

File: core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java]: Removed resourceGroup, can re-add this in elasticity branch

**Comment by dlmarion** [comment]
Closing in favor of #4845

**Comment by keith-turner** [comment]
> Closing in favor of https://github.com/apache/accumulo/pull/4845

#4845 does not include the capability to get RG info for servers in Accumulo.  It would be nice if we could do the following

 * limit listscans and listcompactions output in the shell by RG
 * see RGs in listscan and listcompaction output in the shell

ideally to do the above RGs would need to be better supported in the API.  Opened #4849 to statisfy a subset of this, but a more general approach like this would be better.

**Pull request closed without merging**
